Japanese IME operation panel added

moetuki opened the pull request:
This is my first pull request.
If there is anything that is lacking, I will deal with it sincerely.

If want to use it as a replacement for the original taskbar in a Japanese environment,
Will need a Japanese input control panel.

IMEs are used in languages other than Japanese, e.g. Chinese, Korean.
Countries that use IME may have the same problem.

I have also added an option to select the language of the IME panel,
but currently only Japanese is available.

To verify functionality,
Need to ask users of each language to implement the feature.

Sorry, I'm not familiar with C# and English, but I'll try my best.

moetuki committed:
Japanese IME

Japanese IME status and control

moetuki committed:
Japanese IME

Japanese IME status and control

dremin commented:
Hi, thanks for working on this! It will take some time for me to review, but I wanted to mention one thing I noticed. A lot of the Win32 functions and structs you have defined are already available in `ManagedShell.Interop.NativeMethods`, so you should be able to add `using static ManagedShell.Interop.NativeMethods;` and remove the duplicates.

moetuki committed:
ManagedShell.Interop.NativeMethods

add using static ManagedShell.Interop.NativeMethods; and remove the duplicates.

Corrected data types and removed unused declarations

moetuki commented:
This has been corrected, so thank you for your time.

dremin reviewed (changes requested):
Thanks for patiently waiting, apologies it took me so long! I have been playing with the Japanese IME on both Windows 10 and Windows XP to try and understand how it works.

I left some initial comments below (I will probably have more after the changes 😅). The primary change I'd like to see is having the IME be part of the InputLanguage control, which removes the need for a separate setting per IME, and also removes the need for the "disabled" state.

moetuki commented:
I'm sorry, but I will be working on it this weekend, so I would appreciate your patience.
Thank you for your confirmation.

moetuki committed:
I have corrected the items pointed out

Making the Japanese IME bar part of the language bar
English notation of Japanese IME menu
Left-click to switch between hiragana and direct input
identify Windows folders
Process startup method

moetuki commented on RetroBar/Taskbar.xaml:
```diff
@@ -113,6 +113,7 @@
                                  RelativeSource="{RelativeSource Mode=FindAncestor, AncestorType=Window}" />
                     </StackPanel.Orientation>
                     <controls:NotifyIconList NotificationArea="{Binding NotificationArea}" />
+                    <controls:JapaneseIme VerticalAlignment="Center" />
```
*Making the Japanese IME bar part of the language bar

When Japanese is selected on Japanese Windows, the [JP] display on the language bar is not normally displayed, so
I tried to hide the [JP] display when both the Windows language and the RetroBar selected language are Japanese.

moetuki replied:
I have fixed the issue so that it will not be displayed if there is only one input language, Japanese.

moetuki commented on RetroBar/Languages/English.xaml:
```diff
@@ -138,6 +144,25 @@
+    <s:String x:Key="ime_input_key_kana">Kana Input</s:String>
+    <s:String x:Key="ime_conversion">Conversion Mode</s:String>
+    <s:String x:Key="ime_conversion_general">general</s:String>
+    <s:String x:Key="ime_conversion_none">No Conversion</s:String>
```
*English notation of Japanese IME menu

Thank you for providing the image.
The English notation has been corrected.

moetuki commented on RetroBar/Controls/JapaneseIme.xaml:
*Left-click to switch between hiragana and direct input

Sorry, I had planned to implement this when I started working on it, but I forgot.

moetuki commented on RetroBar/Controls/JapaneseIme.xaml.cs:
```diff
@@ -0,0 +1,800 @@
+            SetForeOtherTopWindow();
+
+            return PrgExec(Prop32Name, Prop64Name, ExecOpt);
+        }
```
*identify Windows folders

Sorry, I forgot to fix it when I tested the element function.

moetuki commented on RetroBar/Controls/JapaneseIme.xaml.cs:
```diff
@@ -0,0 +1,800 @@
+            }
+
+            return RetSts;
+        }
```
*Process startup method

The method has been revised to use ManagedShell.

moetuki commented on RetroBar/Utilities/Settings.cs:
```diff
@@ -375,6 +375,13 @@ public bool HotkeysQuickLaunch
+        {
+            get => _imeShow;
+            set => Set(ref _imeShow, value);
+        }
```
The setting has been removed

moetuki committed:
Removed unused language resources

Removed unused language resources

dremin reviewed (changes requested):
Nice, thanks for the changes @moetuki! After looking closer I have some more feedback.

moetuki committed:
I have corrected the items pointed out

Menu item reference method correction
Modified to use a button in the UserControl
Modified to add/remove the control itself
Forgot to stop the timer, fixed
Timer start timing correction
English resource ime_conversion_general Modification
Language display condition correction

moetuki committed:
Merge branch 'dremin:master' into JpnIME01

moetuki commented on RetroBar/Controls/JapaneseIme.xaml.cs:
```diff
@@ -0,0 +1,795 @@
+            MenuItem Item_InputKana = this.FindName("JapaneseIme_input_key_kana") as MenuItem;
+
+            MenuItem Item_ConversionGeneral = this.FindName("JapaneseIme_conversion_general") as MenuItem;
+            MenuItem Item_ConversionNone = this.FindName("JapaneseIme_conversion_none") as MenuItem;
```
This happened because I wasn't confident about how to reference it in the code.
I have fixed it.

moetuki commented on RetroBar/Controls/JapaneseIme.xaml.cs:
```diff
@@ -0,0 +1,795 @@
+
+            if (this.Visibility == Visibility.Visible)
+            {
+                if (Settings.Instance.Edge == AppBarEdge.Top || Settings.Instance.Edge == AppBarEdge.Bottom)
+                {
+                    this.Width = size;
+                }
+                else if (Settings.Instance.Edge == AppBarEdge.Left || Settings.Instance.Edge == AppBarEdge.Right)
+                {
+                    this.Height = size;
+                }
```
It has been changed to a button within a UserControl.
Without size adjustment, the width of the UserControl itself will change depending on the IME status,
which will result in poor visibility, so we have adjusted the size.

moetuki commented on RetroBar/Controls/JapaneseIme.xaml.cs:
```diff
@@ -0,0 +1,795 @@
+            Settings.Instance.PropertyChanged -= Settings_PropertyChanged;
+
+            _isLoaded = false;
+        }
```
I forgot to process it. I have corrected it.

moetuki commented on RetroBar/Controls/JapaneseIme.xaml.cs:
```diff
@@ -0,0 +1,795 @@
+
+            ImeCheckTimer.Interval = TimeSpan.FromMilliseconds(200);
+            ImeCheckTimer.Tick += ImeChk_Event;
+            ImeCheckTimer.Start();
```
It has been corrected.

moetuki commented:
I apologize for the many corrections.
Thank you for your continued support.

dremin committed:
Theme IME, remove unneeded code

dremin committed:
Remove unneeded code

dremin reviewed (approved):
Thanks for all of your work on this @moetuki!
